### DISCUSSION TITLE
Missing dependencies should not print stack traces

### DISCUSSION THREAD

**Discussion by lokka30** [PaperMC/Paper#10560] [Ideas]
> This issue comment is copy-pasted from [my Spigot thread](https://www.spigotmc.org/threads/missing-dependencies-should-not-print-stack-traces.589039/) which discussed the matter. I decided to also discuss this on Paper's issue tracker to see if you guys are interested in it. Thanks 🙂 

Hi PaperMC community, I hope you're all doing well :)

I would like to bring attention to, and hopefully have resolved, an issue to do with how hard dependencies are handled by Spigot.

[A thread on this identical issue](https://www.spigotmc.org/threads/why-does-spigot-throw-an-error-rather-than-just-print-a-warn-with-missing-dependencies.468326/) was already created by *Andre_601* in 2020, although that thread seems to have dwindled without any official comment on the matter.

When you run a plugin which has 'hard dependencies' (declared via `depend: [...]` in `plugin.yml`), and your server does not have one of those hard dependencies installed, then Spigot will print a rather non-user-friendly message to inform about the missing dependency.

The following console log was taken from [this](https://www.spigotmc.org/threads/error-message-in-console-and-plugins-not-working.530205/) thread which shows you what this thread is about. The linked thread also serves as one of the many examples where a user believes a plugin is at fault (because they see a stack trace), although they have caused it by not following the plugin's installation instructions.

```
[22:31:48 ERROR]: Could not load 'plugins/RankGrantPlus-3.8.0.jar' in folder 'plugins'
org.bukkit.plugin.UnknownDependencyException: Unknown/missing dependency plugins: [Vault]. Please download and install these plugins to run 'RankGrantPlus'.
at org.bukkit.plugin.SimplePluginManager.loadPlugins(SimplePluginManager.java:291) ~[patched_1.17.1.jar:git-Paper-281]
at org.bukkit.craftbukkit.v1_17_R1.CraftServer.loadPlugins(CraftServer.java:410) ~[patched_1.17.1.jar:git-Paper-281]
at net.minecraft.server.dedicated.DedicatedServer.init(DedicatedServer.java:276) ~[patched_1.17.1.jar:git-Paper-281]
at net.minecraft.server.MinecraftServer.x(MinecraftServer.java:1212) ~[patched_1.17.1.jar:git-Paper-281]
at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:319) ~[patched_1.17.1.jar:git-Paper-281]
at java.lang.Thread.run(Thread.java:831) [?:?]
```

To us developers, this stack trace is extremely easy to read - the answer is right at the top. But for others, they see all of the Java jargon packed beneath the thing they should have seen, and assume that 'the plugin has a bug so bad that I can't even run it'. It's not user friendly when combined with the exception information.

To resolve this issue, the exception information should not be printed when this exception is caught by Bukkit's plugin loader. I assume this is quite a simple fix. It could look like this when logged to the console:

> [22:31:48 ERROR]: Could not load 'plugins/RankGrantPlus-3.8.0.jar' in folder 'plugins': You do not have the following dependency plugins installed: [Vault]. Please download and install these plugins to run 'RankGrantPlus'.

To get around the non-user-friendly stack trace, myself among other plugin developers declare hard dependencies as soft dependencies instead, and add logic to our `onEnable` methods to enforce these dependencies, which displays more user-friendly messages.

I'd greatly appreciate to hear your opinion below on this matter. If you maintain a plugin which has hard dependencies, you've likely come across a user who reports this as a bug to you. It's for this reason that I avoid writing plugins that would require any hard dependency.

Thank you everyone, and all the best.
~ lokka30

**Comment by MiniDigger** [comment]
I agree that stack traces are not user friendly, the problem I have with one line error messages is that they will be pretty easily overlooked.
Plugin authors resorting to soft depedencies because of bugs in the current dependency resolution is a different issue, those will hopefully be fixed with the whole paper plugin and lifecycle api stuff.

**Comment by electronicboy** [comment]
We could maybe make a multi line pretty message or something, the user
friendliness of that specific message has been discussed a few times

**Comment by lokka30** [comment]
Here's an idea for a multi-line message:

```
[22:31:48 ERROR]: Could not load 'plugins/ExamplePlugin-1.2.3.jar' in folder 'plugins':
You have not installed the following dependency plugins of 'ExamplePlugin':

  - EssentialsX
  - WorldEdit

Since you have not resolved the dependency plugins of 'ExamplePlugin', it will not be able to load on your server.
Ensure that you have carefully followed the plugin's installation instructions (if provided).
```

**Comment by alfiejfs** [comment]
What would be the ideal of handling this? The exception is created and passed into the logger - to get this behaviour, it would make the most sense to just not create the exception at all?

I'm happy to do this one if it's approved. I think it's a great idea :)

**Comment by alfiejfs** [comment]
![image](https://user-images.githubusercontent.com/42497992/218057689-ef0ba316-00d1-4867-bc07-e77bac797101.png)

Unsure if the spacing is too much (it is one tab), but here is an example of what it could look like.

**Comment by lokka30** [comment]
> What would be the ideal of handling this? The exception is created and passed into the logger - to get this behaviour, it would make the most sense to just not create the exception at all?

I haven't checked the internal code, though my thought (which I should have noted) is:

- before the exception is passed onto the logger, the server will check if the particular exception is `instanceof` some new abstract exception class
- this abstract exception class could be created in case other similar issues to this arise
- the current 'missing dependency' exception in Bukkit would extend that abstract class instead
- back to the first bullet point, if an exception is `instanceof` the abstract one we created, it'll just print the exception message instead of the stack trace alongside it.

Haven't tried this, just brainstorming. 😄 

> Spacing

I think there should be one less vertical space on both sides of the missing dependencies list. I don't mind the tab, in fact, it makes the message stand out more, which is one of the problems we need to resolve for this to be solved. That's personal opinion of course! 👍🏻
